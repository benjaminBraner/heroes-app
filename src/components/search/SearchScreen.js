import { useNavigate, useLocation } from "react-router-dom";
/*para manejar mas facil los query parameters que vienen de la url instalar "query-string" desde
la consola: npm i query-string */
import queryString from "query-string";
import { getHeroByName } from "../../helpers/getHeroByName";
import { useForm } from "../../hooks/useForm"
import { HeroCard } from "../hero/HeroCard";
import { useMemo } from "react";

export const SearchScreen = () => {

     const navigate = useNavigate();

     /*el useLocation() me muestra la URL en la que me encuentro, y si dicha url se actualiza, tambien lo hara useLocation, dentro viene:
     {
          hash: ""
          key: "wyfydxso"
          pathname: "aqui viene la url en la que me encuentro"
          search: "aqui viene el queryString ( empieza con un "?" )"
          state: null 
     } */
     const location = useLocation();
     
     /*la funcion queryString.parse separara queryParameters, por ejemplo si en mi url tengo los queryParameters asi:
     ?q=superman&publisher=dc&nacionalidad=criptoniana

     queryString.parse los ordenaria asi:
     {
          nacionalidad: "criptoniana",
          publisher: "dc",
          q: "superman"
     } 
     
     ahora voy a desestructurar los query params directamente que le voy a mandar en el handleSearch, al principio si no viene nada
     ya sea porque es la primera vez que se renderiza, etc,  sera igual a un string vacio, y  sera el valor inicial del searchText, porque
     si actualizo la pagina manualmente, la url no cambiara y quiero que si se actualiza, el campo de texto "searchText" tenga el valor
     que tiene "q" que el valor de "q" que esta en la url porque como dije antes, al actualizar la pagina, los queryParams no cambian*/
     const { q = "" } = queryString.parse( location.search ); 

     const [ { searchText }, handleInputChange ] = useForm( { searchText: q } );
     
     const heroesFiltered = useMemo( () => getHeroByName(q) ,[q]);

     const handleSearch = (e) => {

          e.preventDefault();
          //al a√±adir este queryString estoy actualizando el URL por lo tanto el componente igual lo hara
          navigate( `?q=${ searchText }` );  //este string es mi queryString se lo identifica por el "?"
     }

     return (
          <div>
               <div className="row">
                    <div className="col-5">
                         <h4 className="mt-3">Search</h4>
                         <hr />

                         
                         <form onSubmit={ handleSearch }>
                              <input
                                   type="text"
                                   placeholder="Search a hero"
                                   className="form-control"
                                   name="searchText"
                                   autoComplete="off"
                                   onChange={handleInputChange}
                                   value={ searchText }
                              />

                              <button
                                   type="submit"
                                   className="btn btn-outline-primary"
                              >
                                   Search...
                              </button>

                         </form>
                    </div>

                    <div className="col-7">
                         <h4>Resultados</h4>
                         <hr />

                         {
                              (q === "") 
                                   ? <div className="alert alert-info">Buscar un heroe</div>       
                                   : ( heroesFiltered.length === 0) 
                                   &&  <div className="alert alert-danger">No hay resultados : {q}</div>         
                         }


                         {
                              heroesFiltered.map( hero => (
                                   <HeroCard 
                                        key={ hero.id }
                                        { ...hero }
                                   />
                              ) )
                         }
                    </div>
               </div>
          </div>
     )
}
