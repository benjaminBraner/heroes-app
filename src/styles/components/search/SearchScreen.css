.SearchScreen {
     display: flex;
     flex-direction: column;
     justify-content: space-between;
     margin: 30px;
     margin-top: 75px;
}
h1 {
     margin: 20px 27px;
     padding: 0.1px;
}

.search-input{
     width: 100%;
     max-width: 400px;
     min-width: 200px;
     margin-right: 8px;
     outline: none;
     border-radius: 4px;
     border: 1px solid gray;
     height: 35px;
     margin-bottom: 8px;
}

.search-btn {
     background-color: #0d6efd;
     border: none;
     padding: 5px;
     border-radius: 4px;
}

.no-found-search{
     background-color: rgb(240, 136, 136);
     padding: 20px;
     width: 500px;
     color: #9b2e2e;
     border-radius: 4px;
}