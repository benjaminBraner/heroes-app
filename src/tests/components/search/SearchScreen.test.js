import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

/*ES IMPORTANTE QUE TENGA EL NOMBRE DE 'mockNavigate' para que funcione,
declaro este mock del useNavigate afuera porque luego lo voy a probar en mi test porque si lo haria de adentro
no habria manera de sacar este mockNavigate que es lo que va a retornar el useNavigate*/
const mockNavigate = jest.fn();

/*yo podria hacer un mock de todo el react-router-dom, pero cosas que estoy usando en mis tests como el
mount() no funcionarian, y ademas solo requiero hacer un mock del useNavigate*/
jest.mock( 'react-router-dom' , () => ({

     /*con esto es como si yo no hiciera nada osea le digo que importe el paquete real del react-router-dom como si no
     pasara nada, esto me va a permitir sobreescribir el useNavigate que es el mock que quiero hacer en realidad*/
     ...jest.requireActual('react-router-dom'),
     useNavigate: () => mockNavigate
}))


describe('Pruebas en <SearchScreen />', () => {

     test('debe mostrarse correctamente con valores por defecto', () => {
          const wrapper = mount( 
               <MemoryRouter initialEntries={ [ '/search' ] }>
                    <SearchScreen /> 
               </MemoryRouter>
          );
          expect( wrapper ).toMatchSnapshot();
          expect( wrapper.find('.alert-info').text().trim() ).toBe('Buscar un heroe');
          
     })


     test('debe mostrar a Batman y el input con el valor del queryString', () => {
          const wrapper = mount( 
               <MemoryRouter initialEntries={ [ '/search?q=batman' ] }>
                    <SearchScreen /> 
               </MemoryRouter>
          );

          expect( wrapper.find( 'input' ).prop( 'value' ) ).toBe( 'batman' );
     })
     

     test('debe mostrar un error si no se encuentra el heroe', () => {
          const wrapper = mount( 
               <MemoryRouter initialEntries={ [ '/search?q=123' ] }>
                    <SearchScreen /> 
               </MemoryRouter>
          );

          expect( wrapper.find( '.alert-danger' ).text().trim() ).toBe( 'No hay resultados : 123' )
     })
     

     test('debe llamar el navigate a la nueva pantalla ', () => {
          const wrapper = mount( 
               <MemoryRouter initialEntries={ [ '/search?q=123' ] }>
                    <SearchScreen /> 
               </MemoryRouter>
          );

          //simulo que hice un cambio en el input con el valor de 'batman'
          wrapper.find( 'input' ).simulate( 'change' , {
               target: {
                    name: 'searchText',
                    value: 'batman'
               }
          })

          /*llamo el submit del formulario que dispara la funcion handleSearch y le proporciono el preventDefault 
          del evento pero esto pareceria que esta bien pero me lanzaria un error porque dentro del handleSearch
          estoy usando la funcion navigate que intenta hacer el cambio de la url pero ese cambio dispara ciertos 
          procedimientos, y el error diria ' An update to MemoryRouter inside a test was not wrapped in act(...)'
          y yo no quiero verificar que el navigate() funcione, yo quiero verificar que sea llamado, y para hacer esta
          prueba debo hacer un mock del useNavigate()*/
          wrapper.find( 'form' ).prop( 'onSubmit' )({
               preventDefault: () => {}
          })

          expect( mockNavigate ).toHaveBeenCalledWith( '?q=batman' );



     })
     
     
})
