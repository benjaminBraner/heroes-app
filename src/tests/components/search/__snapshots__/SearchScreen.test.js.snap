// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe mostrarse correctamente con valores por defecto 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4
        className="mt-3"
      >
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un heroe
      </div>
    </div>
  </div>
</div>
`;
